File: src/data/courses/dsa/trees/binary-tree-core-operations/bt-core-ops/code/java/BtCoreOps.java
// ============================================================
// Binary Tree Core Operations (BST)
// ============================================================

class TreeNode {
    int val; TreeNode left, right;
    TreeNode(int v) { val = v; }
}

class BSTOperations {
    public TreeNode insert(TreeNode root, int val) {
        if (root == null) return new TreeNode(val);
        if (val < root.val) root.left = insert(root.left, val);
        else if (val > root.val) root.right = insert(root.right, val);
        return root;
    }

    public boolean search(TreeNode root, int val) {
        if (root == null) return false;
        if (val == root.val) return true;
        return val < root.val ? search(root.left, val) : search(root.right, val);
    }

    public TreeNode deleteNode(TreeNode root, int val) {
        if (root == null) return null;
        if (val < root.val) root.left = deleteNode(root.left, val);
        else if (val > root.val) root.right = deleteNode(root.right, val);
        else {
            if (root.left == null) return root.right;
            if (root.right == null) return root.left;
            TreeNode succ = root.right;
            while (succ.left != null) succ = succ.left;
            root.val = succ.val;
            root.right = deleteNode(root.right, succ.val);
        }
        return root;
    }
}

File: src/data/courses/dsa/trees/binary-tree-core-operations/bt-core-ops/code/python/bt_core_ops.py
# ============================================================
# Binary Tree Core Operations (BST)
# ============================================================

class TreeNode:
    def __init__(self, val=0):
        self.val = val
        self.left = None
        self.right = None

class BSTOperations:
    def insert(self, root, val):
        if not root: return TreeNode(val)
        if val < root.val: root.left = self.insert(root.left, val)
        elif val > root.val: root.right = self.insert(root.right, val)
        return root

    def search(self, root, val):
        if not root: return False
        if val == root.val: return True
        if val < root.val: return self.search(root.left, val)
        return self.search(root.right, val)

    def delete(self, root, val):
        if not root: return None
        if val < root.val: root.left = self.delete(root.left, val)
        elif val > root.val: root.right = self.delete(root.right, val)
        else:
            if not root.left: return root.right
            if not root.right: return root.left
            # Two children: inorder successor
            succ = root.right
            while succ.left: succ = succ.left
            root.val = succ.val
            root.right = self.delete(root.right, succ.val)
        return root
